### ISSUE TITLE
scipy.stats.mode

### ISSUE THREAD

**Issue by gianlucadetommaso** [google/jax#8978]
It would be great if you could add the JAX equivalent of `scipy.stats.mode`, which is currently unavailable. 

A use case may be a ML classification task with ensembles, where multiple models give different predictions, and we are interested in finding the most common one.

As an example, consider `predictions` to be a two-dimensional DeviceArray of predictions, with 
shape = (number of models, number of data points).

It would be nice if one could compute

`mode_prediction  = jax.scipy.stats.mode(predictions, axis=0)`,

returning the most common prediction for every data point.

**Comment by jakevdp** [comment]
Here's a quick jit-compatible implementation if you want to make use of it:
```python
import jax.numpy as jnp
from jax import vmap

def mode(x, axis=None):
  def _mode(x):
    vals, counts = jnp.unique(x, return_counts=True, size=x.size)
    return vals[jnp.argmax(counts)]
  if axis is None:
    return _mode(x)
  else:
    x = jnp.moveaxis(x, axis, 0)
    return vmap(_mode, in_axes=(1,))(x.reshape(x.shape[0], -1)).reshape(x.shape[1:])

x = jnp.array([1, 1, 2, 2, 2, 3, 4])
print(mode(x))
# 2

y = jnp.array([[1, 1, 2, 3],
               [4, 5, 6, 5]])
print(mode(y, 1))
# [1, 5]
```

**Comment by gianlucadetommaso** [comment]
Great, thanks!

**Comment by jakevdp** [comment]
It's definitely in scope to add this to `jax.scipy.stats`. If anyone wants to tackle it, you could take the example implementation from above, make sure it returns the same values as `scipy.stats`, and add an appropriate test in `tests/scipy_stats_test.py`

**Comment by jayendra13** [comment]
Let me take this up, can anyone suggest a good location/file for this function to put ?
For Scipy it's defined [here](https://github.com/scipy/scipy/blob/v1.8.0rc2/scipy/stats/_mstats_basic.py#L285) and [here](https://github.com/scipy/scipy/blob/v1.8.0rc2/scipy/stats/_stats_py.py#L365).

I am starting this for now by putting it inside the [__init__.py](https://github.com/google/jax/blob/main/jax/scipy/stats/__init__.py)

**Comment by Yash621** [comment]
@gianlucadetommaso  can i work on this ?

**Comment by gianlucadetommaso** [comment]
@Yash621 please go ahead, I haven't had much time to take this up myself.

**Comment by Yash621** [comment]
@gianlucadetommaso can you suggest the location where i should place this function?

**Comment by jakevdp** [comment]
You could put the implementation in a new submodule, maybe something like `jax/_src/scipy/stats/_core.py`

**Comment by elusenji** [comment]
I'd love to jump in and collaborate with @Yash621 on this, if that's okay.

**Comment by parthpatel9414** [comment]
Would love to collaborate with @Yash621, if this issue is still open?

**Comment by jakevdp** [comment]
I think the PR in https://github.com/google/jax/pull/9141 is quite close - it just needs a bit more work before it can be merged.

**Comment by big-c-note** [comment]
@jakevdp @gianlucadetommaso I took a swing at it on a fresh branch (#11475).

(Apologies for the couple of branches. I was tricking the google cla scan to find my cla.). 

Let me know any feedback you have.

**Comment by big-c-note** [comment]
Also for those wanting to contribute - based on any feedback we get here - feel free to open a PR on my branch
